add migration best practice for review (#2515)
AgoraIO/agora_doc_source: en-US/markdown/Video/migration.md

# API settings after SDK migration

> Agora supports upgrading the RTC SDK from v3.x, v4.0.0 Preview, or v4.0.0 Beta to v4.x. For more information on migration steps and differences between versions, see the [Migration guide](https://docs.agora.io/cn /video-call-4.x/migration_guide_windows_ng?platform=Windows).

Reviewer: ```suggestion
Agora supports upgrading the RTC SDK from v3.x, v4.0.0 Preview, or v4.0.0 Beta to v4.x. For more information on migration steps and differences between versions, see the [Migration guide](https://docs.agora.io/cn/video-call-4.x/migration_guide_windows_ng?platform=Windows).
```

> Agora supports upgrading the RTC SDK from v3.x, v4.0.0 Preview, or v4.0.0 Beta to v4.x. For more information on migration steps and differences between versions, see the [Migration guide](https://docs.agora.io/cn /video-call-4.x/migration_guide_windows_ng?platform=Windows).

Reviewer: EN website does not really use "RTC SDK". It's split into Voice SDK and Video SDK, with dedicated doc sets (so ideally two copies of this doc should exist, like there are two copies of the existing migration guide - [video](https://docs.agora.io/en/video-calling/develop/migration-guide?platform=web) and [audio](https://docs.agora.io/en/voice-calling/develop/migration-guide?platform=web)). There are also variables that can be used for product names: 
\<Vg k="ASDK" /> - Voice SDK
\<Vg k="VSDK" /> - Video SDK

Also, the link should probably go to the English migration guide, not Chinese?

This page introduces the scenarios in which the APIs and related implementation codes should be modified to help you smoothly complete migration after you upgrade the SDK.

Unless otherwise specified, all the example APIs on this page are in C++, and the corresponding platform is Windows.

## Using v3.x before migration
If you use features and scenarios described in this section in v3.x, refer to the corresponding example to update the code in your app.

### Single channel

> This section provides sample code for commonly used functions in single channel.

Reviewer: ```suggestion
This section provides sample code for commonly used functions in a single channel.
```

#### Initialize the engine

v3.x

> - Android: [Start Interactive Live Video Streaming]() and [sample project]()

Reviewer: Links to docs like this are not capitalized in the rest of this document. So this one should probably also follow sentence case: "Start interactive live video streaming"

ScreenCaptureSourceInfo info = infos->getSourceInfo(index);
ScreenCaptureParameters capParam;
// Calls startScreenCaptureByDisplayId, startScreenCaptureByWindowId, or startScreenCaptureByScreenRect to share screen
m_rtcEngine->startScreenCaptureByDisplayId((int)info.sourceId, regionRect, capParam);
// Stops screen sharing
m_rtcEngine->stopScreenCapture();
```

v4.x

```C++
SIZE size;
size.cx = 100;
size.cy = 100;
agora::rtc::Rectangle regionRect = { 0,0,0,0 }；
IScreenCaptureSourceList* infos = m_rtcEngine->getScreenCaptureSources(size, size, true);
ScreenCaptureSourceInfo info = infos->getSourceInfo(index); ScreenCaptureParameters capParam;
m_rtcEngine->startScreenCaptureByDisplayId((int)info.sourceId, regionRect, capParam);
ChannelMediaOptions options;
options.publishCameraTrack = false;
// Publishes the screen-sharing stream
options.publishScreenTrack = true;
m_rtcEngine->updateChannelMediaOptions(options);
......
// Stops publishing screen-sharing stream
options.publishScreenTrack = false;
m_rtcEngine->updateChannelMediaOptions(options)
// Stops screen sharing
m_rtcEngine->stopScreenCapture();
```

#### Publish MP4 videos

v3.x 

> In v3.x, this feature requires an independent process and is implemented through MediaPlayer Kit. See ][MediaPlayer Kit (3.x)]().

Reviewer: ```suggestion
In v3.x, this feature requires an independent process and is implemented through MediaPlayer Kit. See [MediaPlayer Kit (3.x)]().
```

- iOS/macOS: `AgoraCIExtension.xcframework`
- Windows: `libagora_segmentation_extension.dll`

#### Implementation

v3.x

- Android: [Share the screen]() and [sample code]()
- iOS: [Share the screen and]() and [sample code]()
- Windows: [Share the screen]() and [sample code]()
- macOS: [Share the screen]() and [sample code]()

v4.x

- Android: [Share the screen]() and [sample code]()
- iOS: [Share the screen]() and [sample code]()
- Windows: [Share the screen]() and [sample code]()
- macOS: [Share the screen]() and [sample code]()

### Custom sources

If you use the custom sources feature in v3.x, such as the custom video source for a third-party image enhancement extension, Agora recommends that you use the more refined and easier-to-use SDK source in v4.x.

v4.x refines the video module in the SDK and adds support for the Texture format on Android. In general, Agora recommends that users use the SDK to capture audio and video sources after migration. In some cases, however, users can continue to use the custom sources feature, such as when they need to connect SDKs from different providers.
Compared to custom sources, the SDK source in v4.x has the following advantages:

> - Easy to use: Users do not need to deal with video capture, device model compatability, and capture profile strategies. v4.x minimizes code complexity for their applications and reduces issues caused by audio and video capture. In addition, using the SDK to capture audio and video sources can better avoid out-of-sync issues.

Reviewer: ```suggestion
- Easy to use: Users do not need to deal with video capture, device model compatability, and capture profile strategies. v4.x minimizes code complexity for their applications and reduces issues caused by audio and video capture. In addition, using the SDK to capture audio and video sources can help avoid out-of-sync issues.
```

- Low cost of issue identification: The SDK supports uploading and collecting data such as resolution, frame rate, and image processing time, and Agora analyzes these data in the background to quickly identify the causes of issues such as frame loss.

> - Low maintenance cost: Agora has R&D team to continuously, iteratively optimize using the SDK to capture audio and video, and QA testing for the SDK source is improved. By contrast, Agora mainly resolves major issues for the custom sources feature and does not provide iterative updates.

Reviewer: ```suggestion
- Low maintenance cost: Agora has an R&D team to continuously, iteratively optimize using the SDK to capture audio and video. QA testing for the SDK source is also consistently improved. By contrast, Agora mainly resolves major issues for the custom sources feature and does not provide iterative updates.
```

- Fast issue resolution: The SDK supports configuring and distributing audio and video capture strategies, and can quickly resolve issues for certain device models.

> - In addition, note that: APIs related to MediaIO have been removed in v4.x; users need to switch to the Push mode.

Reviewer: ```suggestion
- In addition, note that the APIs related to MediaIO have been removed in v4.x; users need to switch to the Push mode.
```

- iOS:
  - [Custom audio source]() and sample projects:
    - [CustomAudioSource]()
    - [ExternalAudio]()
  - [Custom video source]() and sample projects:
    - [CustomVideoSourcePush (Push mode)]()
    - [CustomVideoSourceMediaIO (MediaIO mode)]()
- Windows:
  - [Custom audio source]() and [sample project]()
  - [Custom video source]() and sample projects:
    - [CustomVideoCapture (Push mode)]()
    - [MediaIOCustomVideoCaptrue (MediaIO mode)]()
- macOS:
  - [Custom audio source]() and sample projects:
    - [CustomAudioSource]()
    - [ExternalAudio]()
  - [Custom video source]() and sample projects:
    - [CustomVideoSourcePush (Push mode)]()
    - [CustomVideoSourceMediaIO (MediaIO mode)]()

v4.x

- Android:
  - [Custom video source]() and sample projects:
    - [PushExternalVideo (Push mode)]()
    - [MultiVideoSourceTracks (Video Track method)]()
- iOS:
  - [Custom video source]() and sample code:
    - [CustomVideoSourcePush (Push mode)]()
    - [CustomVideoSourcePushMulti (Video Track method)]()
- Windows:
  - [Custom audio source]() and [sample project]()
- macOS:
  - [Custom audio source]() and [sample project]()

While configuring video capturing, you may ask:

> Q: What are the differences between StopPreview, enableLocalVideo, and muteLocalVideoStream?

Reviewer: ```suggestion
Q: What are the differences between `StopPreview`, `enableLocalVideo`, and `muteLocalVideoStream`?
```

## Using v4.0.0 Preview or v4.0.0 Beta before migration

### Breaking changes

- `AUDIO_SCENARIO_HIGH_DEFINITION` in `AUDIO_SCENARIO_TYPE` is removed. Use its equivalent `AUDIO_SCENARIO_GAME_STREAMING` instead.
- The `option` naming is modified to clarify the control definition for microphone track. `publishAudioTrack` in `ChannelMediaOptions` is changed to `publishMicrophoneTrack`.
- In order to enrich the diversity of multi-stream functions, the video source type parameter is added to the video stream callback events. For example: `onFirstLocalVideoFrame`, `onFirstLocalVideoFramePublished`, `onVideoSizeChanged`, `onVideoPublishStateChanged`, `onLocalVideoStats`, and `onLocalVideoStateChanged`.
- Media Push APIs improvement:
  - To reduce the difficulty of integrating Media Push, this release optimizes the API design of Media Push and improves the handling of network issues within Media Push clients and servers. You can try the optimized Media Push functionality with the following new methods:

>     - `startRtmpStreamWithoutTranscoding`: Starts pushing media streams to a CDN without transcoding. This method works the same as the previous method addPublishStreamUrl(false).

Reviewer: ```suggestion
    - `startRtmpStreamWithoutTranscoding`: Starts pushing media streams to a CDN without transcoding. This method works the same as the previous method `addPublishStreamUrl(false)`.
```

>     - `startRtmpStreamWithTranscoding`: Starts pushing media streams to a CDN and sets the transcoding configuration. This method works the same as calling the previous methods setLiveTranscoding and addPublishStreamUrl(true) in sequence.

Reviewer: ```suggestion
    - `startRtmpStreamWithTranscoding`: Starts pushing media streams to a CDN and sets the transcoding configuration. This method works the same as calling the previous methods `setLiveTranscoding` and `addPublishStreamUrl(true)` in sequence.
```

    - `updateRtmpTranscoding`: Updates the transcoding configuration. This method works the same as the non-first call to the previous method setLiveTranscoding.

>     - `stopRtmpStream`: Stops pushing media streams to a CDN. This method works the same as the previous method removePublishStreamUrl.

Reviewer: ```suggestion
    - `stopRtmpStream`: Stops pushing media streams to a CDN. This method works the same as the previous method `removePublishStreamUrl`.
```

>     - On Android platform, all parameter types in the `onRtmpStreamingStateChanged` callback are changed to int type.

Reviewer: ```suggestion
    - On the Android platform, all parameter types in the `onRtmpStreamingStateChanged` callback are changed to the `int` type.
```

  - This release deprecates three methods for Media Push: `addPublishStreamUrl`, `setLiveTranscoding`, and `removePublishStreamUrl`. Agora recommends that you use the new methods for Media Push and update your code logic by referring to [Media Push]().

> - The `replace` (Booleon type) parameter in `startAudioMixing` is removed. `startAudioMixing` sends the mixing of audio from the microphone and the media player by default. If you need to stop sending audio from the microphone, you can call the following methods as needed depending on whether to continue recording:

Reviewer: ```suggestion
- The `replace` (`Boolean` type) parameter in `startAudioMixing` is removed. `startAudioMixing` sends a mix of audio from the microphone and the media player by default. If you need to stop sending audio from the microphone, you can call the following methods as needed depending on whether to continue recording:
```

>   - Recording required: call `updateChannelMediaOptions` and set p`ublishMicrophoneTrack` to `false`, while the microphone is still on.

Reviewer: ```suggestion
  - Recording required: Call `updateChannelMediaOptions` and set `publishMicrophoneTrack` to `false`, while the microphone is still on.
```

>   - No recording required: call `enableLocalAudio` to turn off the microphone.

Reviewer: ```suggestion
  - No recording required: Call `enableLocalAudio` to turn off the microphone.
```

> - All parameter types in the `onRemoteAudioStateChanged` callback are changed to Int type (Android platform only).

Reviewer: ```suggestion
- All parameter types in the `onRemoteAudioStateChanged` callback are changed to the `int` type (Android platform only).
```

### Behavior changes & added APIs

- `IAudioFrameObserver` is added with these callbacks: `getObservedAudioFramePosition`, `getRecordAudioParams`, `getPlaybackAudioParams`, and `getMixedAudioParams`. `setRecordingAudioFrameParameters` and other series of APIs in v3.x are deprecated.
- Extensions are used for screen sharing. See [Share the screen]().
- The `enableVirtualBackground` method is added with the `SegmentationProperty` and `segProperty` parameters.

> Reference

Reviewer: "Reference" is probably supposed to be a header but not formatted as one?

The following table shows the differences between some common APIs in different versions:

| Common API | `muteLocalAudioStream` | `muteLocalVideoStream` |
| --- | --- | --- |

> | v3.x | The joinChannel method with the options parameter overrides the setting of `muteLocalAudioStream`. | The joinChannel method with the options parameter overrides the setting of `muteLocalVideoStream`.|

Reviewer: ```suggestion
| v3.x | The `joinChannel` method with the `options` parameter overrides the setting of `muteLocalAudioStream`. | The `joinChannel` method with the `options` parameter overrides the setting of `muteLocalVideoStream`.|
```

|v4.0.0 Preview or v4.0.0 Beta | Equivalent to `ChannelMediaOptions.publishAudioTrack`. If the host in the channel calls this method and sets the parameter to `false`, the microphone will be turned on.| Equivalent to `ChannelMediaOptions.publishCameraTrack`. |

> | v4.x	| <ul><li>Calling this method either before or after joining a channel takes effect.</li><li>`muteLocalAudioStream` and `ChannelMediaOptions` are independent of each other: if you set `publishMicrophoneTrack` to `true` and `muteLocalAudioStream` to `true`, the audio stream will not be published.</li></ul> | <ul><li>Calling this method either before or after joining a channel takes effect.</li><li>`muteLocalVideoStream` and `ChannelMediaOptions` are independent of each other: if you set `publishCameraTrack` to `true` and `muteLocalVideoStream` to `true`, the video stream will not be published.</li></ul> |

Reviewer: ```suggestion
| v4.x	| <ul><li>Calling this method either before or after joining a channel takes effect.</li><li>`muteLocalAudioStream` and `ChannelMediaOptions` are independent of each other: if you set `publishMicrophoneTrack` to `true` and `muteLocalAudioStream` to `true`, the audio stream will not be published.</li></ul> | <ul><li>Calling this method either before or after joining a channel takes effect.</li><li>`muteLocalVideoStream` and `ChannelMediaOptions` are independent from each other: If you set `publishCameraTrack` to `true` and `muteLocalVideoStream` to `true`, the video stream will not be published.</li></ul> |
```

> | Common API | setClientRole | startPreview |

Reviewer: ```suggestion
| Common API | `setClientRole` | `startPreview` |
```

| v4.x | <ul><li>`setClientRole` takes precedence over `enableLocalAudio`/`enableLocalVideo`.</li><li>If this method is called before joining a channel, the SDK will first record the status.<ul><li>If you switch to host after joinning the channel, media capture will be enabled and automatically published according to the status of `enableLocalAudio`/`enableLocalVideo`.</li><li>If you switch the user role to audience, media capture will be turned off and publishing will stop.</li></ul></li><li>In a multi-channel scenario, users can switch to host on each channel to publish streams. Stream publishing in the channel is controlled via `updateChannelMediaOptionsEx`.</li></ul> |When the host joins a channel and publishes the video stream captured by the camera, the local preview will be automatically enabled. If `startPreview` is called for your application, you need to call `stopPreview` to close the local preview when leaving the channel. |

> | Common API | enableLocalAudio | enableLocalVideo |

Reviewer: ```suggestion
| Common API | `enableLocalAudio` | `enableLocalVideo` |
```

| --- | --- | --- |
| v3.x | Same as v4.x. | Same as v4.x. |
| v4.0.0 Preview or v4.0.0 Beta	| <ul><li>The microphone will be turned on/off when a host calls this method (either in or outside the channel).</li><li>An error code will be returned when an audience member calls this method.</li><li>If you call `ChannelMediaOptions.publishAudioTrack` in the channel, it takes effect.</li></ul> | <ul><li>The camera will be turned on/off when a host calls this method (either in or outside the channel).</li><li>An error code will be returned when an audience member calls this method.</li><li>If you call `ChannelMediaOptions.publishCameraTrack` in the channel, it takes effect.</li></ul> |

> | v4.x | <ul><li>If you call this method outside the channel, the microphone will not be turned on/off, and the microphone will be turned on/off automatically after you joining the channel.</li><li>If you call this method in a channel, the microphone will be turned on/off.</li> |<ul><li>If you call this method outside the channel, the camera will not be turned on/off, and the camera will be turned on/off automatically after you joining the channel.</li><li>If you call this method in a channel, the camera will be turned on/off.</li> |

Reviewer: ```suggestion
| v4.x | <ul><li>If you call this method outside the channel, the microphone will not be turned on/off, and the microphone will be turned on/off automatically after you joining the channel.</li><li>If you call this method in a channel, the microphone will be turned on/off.</li> |<ul><li>If you call this method outside the channel, the camera will not be turned on/off, and the camera will be turned on/off automatically after you join the channel.</li><li>If you call this method in a channel, the camera will be turned on/off.</li> |
```
